packages/vite/src/node/optimizer/index.ts

```diff
@@ -240,48 +247,51 @@ export interface DepOptimizationMetadata {
  * Scan and optimize dependencies within a project.
  * Used by Vite CLI when running `vite optimize`.
  */
+// TODO: do we need this? It is exposed for the CLI command `vite optimize`
+
 export async function optimizeDeps(
   config: ResolvedConfig,
   force = config.optimizeDeps.force,
   asCommand = false,
 ): Promise<DepOptimizationMetadata> {
   const log = asCommand ? config.logger.info : debug
-  const ssr = false
+
+  // TODO: Could we avoid the creation of a DevEnvironment moving the plugin resolving to
+  // the Environment base class?
+  const environment = new ScanEnvironment('client', config)
+  await environment.init()
 
   const cachedMetadata = await loadCachedDepOptimizationMetadata(
-    config,
-    ssr,
+    environment,
     force,
     asCommand,
   )
   if (cachedMetadata) {
     return cachedMetadata
   }
 
-  const deps = await discoverProjectDependencies(config).result
+  const deps = await discoverProjectDependencies(environment).result
 
-  await addManuallyIncludedOptimizeDeps(deps, config, ssr)
+  await addManuallyIncludedOptimizeDeps(environment, deps)
 
   const depsString = depsLogString(Object.keys(deps))
   log?.(colors.green(`Optimizing dependencies:\n  ${depsString}`))
 
-  const depsInfo = toDiscoveredDependencies(config, deps, ssr)
+  const depsInfo = toDiscoveredDependencies(environment, deps)
 
-  const result = await runOptimizeDeps(config, depsInfo, ssr).result
+  const result = await runOptimizeDeps(environment, depsInfo).result
 
   await result.commit()
 
   return result.metadata
 }
 
-export async function optimizeServerSsrDeps(
-  config: ResolvedConfig,
+export async function optimizeExplicitEnvironmentDeps(
+  environment: Environment,
 ): Promise<DepOptimizationMetadata> {
-  const ssr = true
   const cachedMetadata = await loadCachedDepOptimizationMetadata(
-    config,
-    ssr,
-    config.optimizeDeps.force,
+    environment,
+    environment.config.dev.optimizeDeps.force ?? false, // TODO: should force be per-environment?
```

Comment:
The TODO is outdated. force and entries are already per-environment (as we moved the whole DepOptimizationOptions from root and `ssr.` to `dev.`). So it is now simpler to have it per-environment. I'll remove the TODO for now, and we can revisit this before the release if needed.

As a note, I see many projects adding `force: true` because they find a bug, somehow it isn't exposed during cold start, and then they end up in cold starts forever. This option may be hurting more than what it helps. Maybe it could be made internal or something to discourage its use.

Reply:
If it's already per-environment, I think it's fine as-is 👍 

Yeah, I think it makes sense to change `force` as an internal option.